src/fields/field_fe.cc

```diff
@@ -223,6 +284,59 @@ void FieldFE<spacedim, Value>::value_list (const Armor::array &point_list, const
 
 
 
+template <int spacedim, class Value>
+void FieldFE<spacedim, Value>::cache_update(FieldValueCache<typename Value::element_type> &data_cache,
+		ElementCacheMap &cache_map, unsigned int region_idx)
+{
+    ASSERT( !boundary_dofs_ ).error("boundary field NOT supported!!\n");
+    std::shared_ptr<EvalPoints> eval_points = cache_map.eval_points();
+    Armor::ArmaMat<typename Value::element_type, Value::NRows_, Value::NCols_> mat_value;
+
+    if (fe_values_[0].dim() == -1) {
+        // initialize FEValues objects (when first using)
+        this->init_fe_value<1>(eval_points);
+        this->init_fe_value<2>(eval_points);
+        this->init_fe_value<3>(eval_points);
+    }
+
+    auto update_cache_data = cache_map.update_cache_data();
+    std::unordered_map<unsigned int, typename ElementCacheMap::RegionData>::iterator reg_elm_it =
+            update_cache_data.region_cache_indices_map_.find(region_idx);
+
+    for (unsigned int i_elm=0; i_elm<reg_elm_it->second.element_set_.size(); ++i_elm) {
+        unsigned int elm_idx = reg_elm_it->second.elm_indices_[i_elm];
+    	ElementAccessor<spacedim> elm(const_cast<const Mesh *>(dh_->mesh()), elm_idx);
+        fe_values_[elm.dim()-1].reinit( elm );
+
+        DHCellAccessor cell = dh_->cell_accessor_from_element( elm_idx );
+        LocDofVec loc_dofs = cell.get_loc_dof_indices();
+
+        for (unsigned int i_ep=0; i_ep<eval_points->max_size(); ++i_ep) { // i_eval_point
+            //DHCellAccessor cache_cell = cache_map(cell);
+            int field_cache_idx = cache_map.get_field_value_cache_index(cache_map(cell).element_cache_index(), i_ep);
+            if (field_cache_idx < 0) continue; // skip
+            mat_value.fill(0.0);
+    		for (unsigned int i_dof=0; i_dof<loc_dofs.n_elem; i_dof++) {
+    		    mat_value += data_vec_[loc_dofs[i_dof]]
```

Comment:
This step is basically matrix-vector multiplication. with FeValues matrix of shape n_values * n_dofs, multiplied by
the dof value vector. See implementation of FeValues::xyz_view and try use here shape_value_component directly.
That way we can avoid EvalShapeHandler and open space for more optimization later on.

Reply:
consult with @jstebel  transpose in tensor case

Reply:
1. @jstebel Check how tensor elements are indexed (column-wise/row-wise)
2. @dflanderka Check how we treat GMSH tensors in IO operations.

Reply:
GMSH treats tensors in "row-wise" order. I.e. values "1 2 3 4 5 6 7 8 9" correspond to the tensor:
1 2 3
4 5 6
7 8 9

Simple example: In the following MSH file, plugin Eigenvectors produces vectors in X and Z axis, which means that the values " 1 1 0 0 1 0 0 0 1" in the last but one line represent the tensor:
1 1 0
0 1 0
0 0 1

$MeshFormat
2.2 0 8
$EndMeshFormat
$PhysicalNames
1
3       100     "tetra"
$EndPhysicalNames
$Nodes
4
1 0 0 0
2 1 0 0
3 0 1 0
4 0 0 1
$EndNodes
$Elements
1
1 4 2 100 100 1 2 3 4
$EndElements

$ElementData
1
"tensor"
1
0.0
3
0
9
1
1 1 1 0 0 1 0 0 0 1
$EndElementData

Reply:
Data object in FieldFE (VectorMPI) is in "row-wise" order too. That is for tensor:
1 2 3
4 5 6
7 8 9
Values are stored in order "1 2 3 4 5 6 7 8 9".

Only difference is in armadillo "inner" data represantation  that is stored in "col-wise". If we get mat.memptr() and iterate through items, result is "1 4 7 2 5 8 3 6 9". But all other (print to output, item on position i,j) result correct values of original tensor.

Both - code in FieldFE and unit test (field_evaluate_fe_test) - is correct. There si two ways how to initialize matrix with initializer list:
arma::mat33 A = {1, 4, 7, 2, 5, 8, 3, 6, 9};
arma::mat33 B = { {1, 2, 3}, {4, 5, 6}, {7, 8, 9} }; // A == B
Unit test uses first initialization and therefore it looks like an error.

Reply:
Proposal of replace of EvalShapeHandler is in 584d1a6.